mesonbuild/compilers.py

```diff
@@ -430,6 +430,17 @@ def _can_dedup(cls, arg):
         to recursively search for symbols in the libraries. This is not needed
         with other linkers.
         '''
+
+        # A Standalone -D must never be deduplicated because it is defined
+        # by what comes _after_ it. Thus dedupping this:
+        # -D FOO -D BAR
+        # would yield either
+        # -D FOO BAR
+        # or
+        # FOO -D BAR
+        # both of which are invalid.
+        if arg == '-D' or arg == '/D':
+            return 0
```

Comment:
`arg.startswith()` will still match, so we have to also check that `len(arg) > 2`